### PR TITLE
Adding Open Images v7 to the zoo

### PR DESCRIPTION
**Pull request by jacobmarks**
## What changes are proposed in this pull request?

Added open images v7 support to our model zoo. It can be loaded from the model zoo with 
`dataset = foz.load_zoo_dataset('open-images-v7')`. The only difference from open images v6 is 
support for "point" labels, which are incorporated via the Keypoints Label class. Support for open images v6 is unchanged, despite the backend details being shuffled around.

When the dataset is loaded with the "points" label, each point is represented by a separate Keypoint. I chose this approach because the points in the open images dataset have non-uniform numbers of yes/no votes and methods of generation.

The one detail I'm not sure about is that currently, `dataset.info` is carrying along a `point_classes` key. This is because there were points with class labels that were _not_ in the `classes_map`, but were in the point_classes metadata. Unsure if we should pass this through to the user, or just hide it internally.

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [x] Core: Core `fiftyone` Python library changes
-   [x] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

**Comment by jacobmarks** [comment]
@brimoor just updated the Open Images tutorial to get the bare bones point labels incorporated and explained in there. The tutorial threw an error because I don't have tensorflow in this particular conda env, but I didn't change any of the code that goes into that part of the tutorial. 

Let me know if you think we should include the type of point IoU evaluation they introduced in the paper https://arxiv.org/pdf/2210.14142v2.pdf or if we think this is sufficient

File: fiftyone/zoo/datasets/base.py [hunk 1]
```diff
@@ -2468,11 +2468,203 @@ def supported_splits(self):
...
+    @property
+    def name(self):
+        return "open-images-v7"
+
+    @property
+    def tags(self):
+        return (
+            "image",
+            "detection",
+            "point",
```
Comment (brimoor) [inline fiftyone/zoo/datasets/base.py:2651]: keypoint?

File: fiftyone/zoo/datasets/base.py [hunk 2]
```diff
@@ -2468,11 +2468,203 @@ def supported_splits(self):
...
+
+        return dataset_type, num_samples, classes
+
+
+class OpenImagesV7Dataset(FiftyOneDataset):
+    """Open Images V7 is a dataset of ~9 million images, roughly 2 million of
+    which are annotated and available via this zoo dataset.
+
+    The dataset contains annotations for classification, detection,
+    segmentation, and visual relationship tasks for the 600 boxable object
```
Comment (brimoor) [inline fiftyone/zoo/datasets/base.py:2499]: Update to mention keypoints?

File: fiftyone/zoo/datasets/base.py [hunk 3]
```diff
@@ -2468,11 +2468,203 @@ def supported_splits(self):
...
+    classes.
+
+    This dataset supports partial downloads:
+
+    -   You can specify subsets of data to download via the``label_types``,
+        ``classes``, ``attrs``, and ``max_samples`` parameters
+    -   You can specify specific images to load via the ``image_ids`` parameter
+
+    See :ref:`this page <dataset-zoo-open-images-v6>` for more information
```
Comment (brimoor) [inline fiftyone/zoo/datasets/base.py:2508]: Unfortunately, when adding new zoo datasets, you must manually add a new entry to the docs here:
https://github.com/voxel51/fiftyone/blob/develop/docs/source/user_guide/dataset_zoo/datasets.rst

It is largely copy-paste from this docstring.

After doing so, you can update this ref to `dataset-zoo-open-images-v7`.

File: fiftyone/utils/openimages.py [hunk 4]
```diff
@@ -93,9 +95,20 @@ def __init__(
         seed=None,
         max_samples=None,
     ):
-        _label_types = _parse_label_types(label_types)
+
+        if label_types == None:
```
[the remark this answers is not in the document]
Comment (brimoor) [reply fiftyone/utils/openimages.py:99]: Alternatively, it would be cleaner to refactor this logic into the `_parse_label_types` function.

You could still check for `points.csv` in this constructor as a heuristic for `version`, but then just pass `_parse_label_types(label_types, version)` where the rest of the logic is applied.

**Review by brimoor — COMMENTED** [review]
@jacobmarks looking good! Just left a few comments that are largely stylistic.

To answer your question: no, I don't think we need to implement keypoint evaluation in this PR; that can come later if requested by users.

File: fiftyone/utils/openimages.py [hunk 5]
```diff
@@ -984,7 +1087,7 @@ def _get_all_label_data(
     all_attrs_ids = set(image_ids)
     any_attrs_ids = set()
 
-    _label_types = _parse_label_types(label_types)
+    _label_types = label_types
```
[the remark this answers is not in the document]
Comment (jacobmarks) [reply fiftyone/utils/openimages.py:1090]: As is, it works as you would expect/desire. When `label_types=None`, it assumes all label types that are allowed for that version should be included.

File: fiftyone/utils/openimages.py [hunk 6]
```diff
@@ -1308,12 +1456,11 @@ def _download(
             max_samples,
         )
 
-    if "segmentations" in _parse_label_types(label_types):
+    if "segmentations" in label_types:
```
[the remark this answers is not in the document]
Comment (jacobmarks) [reply fiftyone/utils/openimages.py:1459]: This is dealt with in what is passed in to `_setup()`. I didn't change the default away from `None` because I wasn't sure about our conventions

**Comment by jacobmarks** [comment]
@brimoor just updated. Made all of the suggested edits/improvements except for the `_parse_label_types()` direction. Docs build as desired

File: fiftyone/utils/openimages.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ehofesmann) [reply fiftyone/utils/openimages.py:99]: I think the best approach here would be to check on disk what label types are available and parse the `version` automatically based on the presence of a `points.csv`. This way the `version` doesn't need to be provided explicitly, but all of the other methods that need the `version` will have it available (see my other comments).

It should also raise an informative error if a label type was specified that is not available.

File: fiftyone/utils/openimages.py [hunk 7]
```diff
@@ -596,9 +675,9 @@ def _setup(
     seed=None,
     download=False,
 ):
-    did_download = False
-    _label_types = _parse_label_types(label_types)
 
+    did_download = False
+    _label_types = label_types
```
Comment (ehofesmann) [inline fiftyone/utils/openimages.py:680]: This should use `_parse_label_types()` and have the version passed into `_setup(..., version, ...)`.

File: fiftyone/utils/openimages.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ehofesmann) [reply fiftyone/utils/openimages.py:1088]: When `label_types=None`, this will raise an error when it gets to `if "classifications" in _label_types`. Following my other comments, I'd suggest updating the implementation so `label_types=None` is still allowed by adding a required `version` argument to this method.

**Review by ehofesmann — CHANGES_REQUESTED** [review]
This is great! It works really well!
I left two minor comments and then a few around the same code structure topic Brian was commenting on regarding the parsing of `label_types`.

In general, any of the functions that have a `label_types=None` keyword argument need to be able to work when `label_types=None`. To do that, I would go back to using `_parse_label_types()` everywhere, but use your new method that also requires the version, then just make sure that any method that calls `_parse_label_types()` also has access to the version. It seems that this is possible for all methods (like `_download(...)`) since they get called by methods that have the `version` information.

The only place where that is not the case is the `OpenImagesDatasetImporter` which does not take a `version` as input. This has a bigger problem though in that if you provide label types that are not downloaded, it will error out, even in the existing version. So there needs to be a method written that parses the available label types anyway. With that, we then need to compare the provided `label_types` with the available label types, and if the provided label types is `None`, we just use all available ones. If there are `label_types` specified that are not available, raise an error. This error would currently pop up here: https://github.com/voxel51/fiftyone/blob/eaf0e124f7b5622564febdd4bbde85c75f7a14b2/fiftyone/utils/openimages.py#L1119


Not for this PR, but I also opened this issue since I feel like you should be able to filter by the labels of these Keypoints in the App: https://github.com/voxel51/fiftyone/issues/2507

**Comment by ehofesmann** [comment]
Hmm, renaming the `OpenImagesV6Dataset` type will break existing `open-images-v6` datasets people have since we store this in the `~/fiftyone/open-images-v6/info.json`: 
```
{                                                                                                                                                                                                                                                                                  
    "name": "open-images-v6",                                                                                                                                                                                                                                                      
    "zoo_dataset": "fiftyone.zoo.datasets.base.OpenImagesV6Dataset",                                                                                                                                                                                                               
    "dataset_type": "fiftyone.types.dataset_types.OpenImagesV6Dataset",                                                                                                                                                                                                            
    "num_samples": 1676,                                                                                                                                                                                                                                                           
    "downloaded_splits": {                                                                                                                                                                                                                                                         
        "validation": {                                                                                                                                                                                                                                                            
            "split": "validation",                                                                                                                                                                                                                                                 
            "num_samples": 250                                                                                                                                                                                                                                                     
        },                                                                                                                                                                                                                                                                         
        "test": {                                                                                                                                                                                                                                                                  
            "split": "test",                                                                                                                                                                                                                                                       
            "num_samples": 253                                                                                                                                                                                                                                                     
        },                                                                                                                                                                                                                                                                         
        "train": {                                                                                                                                                                                                                                                                 
            "split": "train",                                                                                                                                                                                                                                                      
            "num_samples": 1173                                                                                                                                                                                                                                                    
        }                                                                                                                                                                                                                                                                          
    },                                                                                                                                                                                                                                                                             
    "classes": [                                                                                                                                                                                                                                                                   
        "Accordion",                                                                                                                                                                                                                                                               
        "Adhesive tape",                                                                                                                                                                                                                                                           
        "Aircraft",         
...
```

So it will try to load that dataset type and not find it.

**Comment by brimoor** [comment]
> Hmm, renaming the `OpenImagesV6Dataset` type will break existing `open-images-v6` datasets people have since we store this in the `~/fiftyone/open-images-v6/info.json`:
> So it will try to load that dataset type and not find it.

@jacobmarks FYI there actually a place in the code where we explicitly account for historical name changes like this. You can just add a check at the place linked below that will replace `OpenImagesV6Dataset` with `OpenImagesDataset` in the dataset type:
https://github.com/voxel51/fiftyone/blob/f6fea3686c032a942f182ff835651dd19ab26b2e/fiftyone/zoo/datasets/__init__.py#L1294

**Comment by jacobmarks** [comment]
> 

@brimoor which approach would you prefer? I have it working with `OpenImagesV6Dataset` and `OpenImagesV7Dataset`, but I'm happy to do the historical name change you suggested. Either way

**Comment by brimoor** [comment]
> > 
> 
> @brimoor which approach would you prefer? I have it working with `OpenImagesV6Dataset` and `OpenImagesV7Dataset`, but I'm happy to do the historical name change you suggested. Either way

Either is fine. Mostly stylistic. A name like `OpenImagesDataset` implies that we intend to support loading all future versions of the dataset using that type. That *is* how `COCODetectionDataset` works, for example. Loads both COCO-2014 and COCO-2017 datasets.

**Comment by jacobmarks** [comment]
> > > 
> > 
> > 
> > @brimoor which approach would you prefer? I have it working with `OpenImagesV6Dataset` and `OpenImagesV7Dataset`, but I'm happy to do the historical name change you suggested. Either way
> 
> Either is fine. Mostly stylistic. A name like `OpenImagesDataset` implies that we intend to support loading all future versions of the dataset using that type. That _is_ how `COCODetectionDataset` works, for example. Loads both COCO-2014 and COCO-2017 datasets.

@brimoor in general I think I like your suggestion of having a future-facing `OpenImagesDataset`, but given the issue with version parsing and how the solution we had involved just looking for "points.csv", maybe distinguishing `OpenImagesDatasetV6` and `OpenImagesDatasetV7` is the right call?

**Review by ehofesmann — APPROVED** [review]
Nice! I'm on board with the changes you made to support `OpenImagesv6`. This all looks good to me, great work on this!
